### DISCUSSION TITLE
Efficient Model Training and Incorporating Feedback: Exploring Freezing Layers

### DISCUSSION THREAD

**Discussion by kshitijrajsharma** [devglobalpartners/ramp-code#7] [Q&A] [answered]
Hi everyone,
I’m writing to you from the [fAIr](https://github.com/hotosm/fAIr) project, which is a free and [open source AI project](https://www.hotosm.org/tech-blog/hot-tech-talks-fair/) initiated by the Humanitarian OpenStreetMap Team (HOTOSM). In fAIr, we are using Ramp as a base model and fine tune new models baked on a customized new training dataset to create local models for specific area/projects.
We are currently training and fitting our base models, but we are facing a challenge when it comes to incorporating feedback. Feedback is usually smaller in number, sometimes only 10 or 20 images. Whenever we receive feedback, we want to apply it (re-train again) to the published checkpoint of the model.

One approach that we have considered is to combine the feedback with the training data and retrain the model again. However, this would take a lot of time, and we are looking for a more efficient way to do this. Additionally, we are expecting an overfitting issue as the model knows already the original dataset
I am thinking of freezing the layers of the training model and updating them only on the new images or feedback that we receive. Keras has a feature that allows you to do this, and here is a snapshot of the example code that i am talking about
```
if cfg[“feedback”][“freeze_layers”] :
    for layer in the_model.layers:
        layer.trainable = False # 
```
I am seeking expert feedback on how this will play with Ramp. If the RAMP team has already tested this approach (freezing layers), I would love to know what you think and how was the performance . What do you think is the best way to fit the model to new learnings or feedback in an appropriate way?
Reference : https://blog.keras.io/building-powerful-image-classification-models-using-very-little-data.html 
https://keras.io/guides/transfer_learning/

**Comment by cjohnstonconsulting** [comment, marked as answer]
Hi Kshitij,

What you propose is almost exactly the use case we envisioned for Ramp. Start with a baseline model, then further train it using new data. I would certainly not freeze and then fine-tune using the same data used to train the model in the first place -- you'd overfit to the original training set, as you suspected. But freezing lower layers, and then fine-tuning using new data, should work as you expect. 

We didn't have time in the project to finish testing this fine-tuning workflow. However, I can tell you what we discovered: the baseline model was trained largely on dry AOIs, so it was going to require a lot of new data to fine tune it to wet or heavily vegetated AOIs. Fine tuning the dry land baseline model to wet/veg AOIs took longer and used more data than I expected. 

You could freeze and then fine-tune with just new data, representative of the new area that you'd be studying. You'd need quite a bit of that, but not as much as was required to train the baseline model. 

How much 'feedback data' do you have? And what sort of AOIs are you working with? 

Here's a thought to consider -- if you're going to be doing a lot of wet/veg AOIs, you might consider training a new baseline ramp model on a data set from a wet/veg area. Then you can fine tune starting with that baseline model instead of the dry land one. I think things would go much more smoothly. (This was the direction we were thinking of, for next steps, by the end of the Ramp development project). 

Cheers! Carolyn
